# Coefficient B10

NS final-interval case: `i=N, j=N`, with $n=N-1$. This is a new derivation in the author’s coefficient layout, using **p** for normalized power. It replaces the corresponding final-interval placement; earlier intervals remain unchanged. SN transposes NS.

## Definition

Set $\chi_1=\chi_{N-1}$, $\chi_2=\chi_N$ and $\Delta\chi=\chi_2-\chi_1$.
The density hats are $L(\chi)=(\chi_2-\chi)/\Delta\chi$ and $R(\chi)=(\chi-\chi_1)/\Delta\chi$.
For $\chi_1\leq\chi\leq\chi_2$, the source lower limit is the evaluation point:

$$F(\chi)=\int_\chi^{\chi_2}\frac{\chi_2-u}{\Delta\chi}\frac{u-\chi}{u}\,du,
\qquad H(\chi)=\int_\chi^{\chi_2}\frac{u-\chi_1}{\Delta\chi}\frac{u-\chi}{u}\,du.$$

There is no source half beyond $\chi_N$. Below $\chi_1$, the rising source instead has its full support $[\chi_1,\chi_2]$. At $\chi_1>0$ the value and first derivative match; both partial source integrals vanish at $\chi_2$.


## Change of variables

$$t=\frac{\chi}{\chi_2},\quad a=1-\frac{\chi_1}{\chi_2},\quad
p=1-\frac{P_1}{P_2},\quad z=1-\frac{1+z_1}{1+z_2}.$$

For an ordinary interval $0<a<1$, $1-a\leq t\leq1$, and $p,z$ are real. The normalized power notation presumes $P_2\neq0$; an endpoint-linear equivalent below removes that restriction. $P_1,P_2$ may be signed effective component power.

$$Q(t)=1-\frac{p}{a}(1-t),\qquad Z(t)=1-\frac{z}{a}(1-t).$$
The integrated source functions are $F(\chi)=\chi_2F_a(t)$ and $H(\chi)=\chi_2H_a(t)$:

$$F_a(t)=\frac{1-t^2}{2a}+\frac{t\log t}{a},$$
$$H_a(t)=\frac{(1-t)^2}{2a}-\frac{(1-a)(1-t)}a-\frac{(1-a)t\log t}a.$$
The continuous endpoint value is $t\log t\to0$ as $t\to0^+$. Physical lensing amplitudes are applied by the component’s existing factor/effective-power convention, exactly once; these radial basis coefficients introduce no additional amplitude.


In [1]:
import sympy as sp
from IPython.display import display

t, a, p, z, b = sp.symbols('t a p z b', real=True)
Q = 1-p*(1-t)/a
Z = 1-z*(1-t)/a
F = ((1-t*t)/2+t*sp.log(t))/a
H = ((1-t)**2/2-(1-a)*(1-t)-(1-a)*t*sp.log(t))/a
u = sp.symbols('u', positive=True)
# Moving-limit source integration, expressed in the same normalized units.
F_PRIMITIVE = sp.integrate((1-u)*(u-t)/(a*u), u)
H_PRIMITIVE = sp.integrate((u-1+a)*(u-t)/(a*u), u)
assert sp.simplify(F_PRIMITIVE.subs(u,1)-F_PRIMITIVE.subs(u,t)-F) == 0
assert sp.simplify(H_PRIMITIVE.subs(u,1)-H_PRIMITIVE.subs(u,t)-H) == 0
display(F, H)


(-t**2/2 + t*log(t) + 1/2)/a

(-t*(1 - a)*log(t) - (1 - a)*(1 - t) + (1 - t)**2/2)/a

## Ordinary coefficient $I_{10}$

$$I_{10}=\int_{1-a}^1 \frac{\left(1 - \frac{p \left(1 - t\right)}{a}\right) \left(1 - \frac{z \left(1 - t\right)}{a}\right) \left(a + t - 1\right) \left(- t \left(1 - a\right) \log{\left(t \right)} - \left(1 - a\right) \left(1 - t\right) + \frac{\left(1 - t\right)^{2}}{2}\right)}{a^{2} t}\,dt.$$

The dimensional coefficient is $B_{10}=\chi_2 P_2(1+z_2) I_{10}$. Expand into powers of $t$ and $\log t$, then integrate each moment. For $k\neq-1$, the primitive of $t^k(\log t)^r$ is obtained by integration by parts; for $k=-1$, use $(\log t)^{r+1}/(r+1)$.

In [2]:
def integrate_moments(expression, lower):
    """Integrate every t**n log(t)**r term exactly on [lower, 1]."""
    result = 0
    for term in sp.Add.make_args(sp.expand(expression)):
        powers = term.as_powers_dict()
        n = int(powers.get(t, 0))
        r = int(powers.get(sp.log(t), 0))
        coefficient = term / t**n / sp.log(t)**r
        if n == -1:
            result -= coefficient*sp.log(lower)**(r+1)/sp.Integer(r+1)
        else:
            k = sp.Integer(n+1)
            primitive = t**k*sum((-1)**j*sp.factorial(r)/sp.factorial(r-j)
                                *sp.log(t)**(r-j)/k**(j+1) for j in range(r+1))
            boundary = 0 if lower == 0 else primitive.subs(t, lower)
            result += coefficient*(primitive.subs(t, 1)-boundary)
    return sp.factor(result)

INTEGRAND = H*(t-1+a)/(a*t)*Q*Z
EXPANDED = sp.expand(INTEGRAND)
I10 = integrate_moments(EXPANDED, 1-a)
display(EXPANDED, I10)


log(t) - 1 + 1/t + p*t*log(t)/a - p*t/a - p*log(t)/a + 2*p/a - p/(a*t) + t*z*log(t)/a - t*z/a + t*log(t)/a - t/(2*a) - z*log(t)/a + 2*z/a - 2*log(t)/a + 2/a - z/(a*t) - 3/(2*a*t) + p*t**2*z*log(t)/a**2 - p*t**2*z/a**2 + p*t**2*log(t)/a**2 - p*t**2/(2*a**2) - 2*p*t*z*log(t)/a**2 + 3*p*t*z/a**2 - 3*p*t*log(t)/a**2 + 5*p*t/(2*a**2) + p*z*log(t)/a**2 - 3*p*z/a**2 + 2*p*log(t)/a**2 - 7*p/(2*a**2) + p*z/(a**2*t) + 3*p/(2*a**2*t) + t**2*z*log(t)/a**2 - t**2*z/(2*a**2) + t**2/(2*a**2) - 3*t*z*log(t)/a**2 + 5*t*z/(2*a**2) - t*log(t)/a**2 - t/(2*a**2) + 2*z*log(t)/a**2 - 7*z/(2*a**2) + log(t)/a**2 - 1/(2*a**2) + 3*z/(2*a**2*t) + 1/(2*a**2*t) + p*t**3*z*log(t)/a**3 - p*t**3*z/(2*a**3) + p*t**3/(2*a**3) - 4*p*t**2*z*log(t)/a**3 + 3*p*t**2*z/a**3 - p*t**2*log(t)/a**3 - p*t**2/a**3 + 5*p*t*z*log(t)/a**3 - 6*p*t*z/a**3 + 2*p*t*log(t)/a**3 - 2*p*z*log(t)/a**3 + 5*p*z/a**3 - p*log(t)/a**3 + p/a**3 - 3*p*z/(2*a**3*t) - p/(2*a**3*t) + t**3*z/(2*a**3) - t**2*z*log(t)/a**3 - t**2*z/a**3 + 2*t*z*log(t)/a**3

(60*a**5*p*z*log(1 - a) - 113*a**5*p*z - 120*a**5*p*log(1 - a) + 250*a**5*p - 120*a**5*z*log(1 - a) + 250*a**5*z + 360*a**5*log(1 - a) - 960*a**5 - 60*a**4*p*z*log(1 - a) - 115*a**4*p*z + 120*a**4*p*log(1 - a) + 440*a**4*p + 120*a**4*z*log(1 - a) + 440*a**4*z - 1800*a**4*log(1 - a) + 1800*a**4 - 390*a**3*p*z + 1080*a**3*p*log(1 - a) - 1380*a**3*p + 1080*a**3*z*log(1 - a) - 1380*a**3*z + 2160*a**3*log(1 - a) - 720*a**3 - 960*a**2*p*z*log(1 - a) + 1230*a**2*p*z - 1680*a**2*p*log(1 - a) + 600*a**2*p - 1680*a**2*z*log(1 - a) + 600*a**2*z - 720*a**2*log(1 - a) + 1500*a*p*z*log(1 - a) - 540*a*p*z + 600*a*p*log(1 - a) + 600*a*z*log(1 - a) - 540*p*z*log(1 - a))/(720*a**4)

### Result

$$I_{10}=\frac{60 a^{5} p z \log{\left(1 - a \right)} - 113 a^{5} p z - 120 a^{5} p \log{\left(1 - a \right)} + 250 a^{5} p - 120 a^{5} z \log{\left(1 - a \right)} + 250 a^{5} z + 360 a^{5} \log{\left(1 - a \right)} - 960 a^{5} - 60 a^{4} p z \log{\left(1 - a \right)} - 115 a^{4} p z + 120 a^{4} p \log{\left(1 - a \right)} + 440 a^{4} p + 120 a^{4} z \log{\left(1 - a \right)} + 440 a^{4} z - 1800 a^{4} \log{\left(1 - a \right)} + 1800 a^{4} - 390 a^{3} p z + 1080 a^{3} p \log{\left(1 - a \right)} - 1380 a^{3} p + 1080 a^{3} z \log{\left(1 - a \right)} - 1380 a^{3} z + 2160 a^{3} \log{\left(1 - a \right)} - 720 a^{3} - 960 a^{2} p z \log{\left(1 - a \right)} + 1230 a^{2} p z - 1680 a^{2} p \log{\left(1 - a \right)} + 600 a^{2} p - 1680 a^{2} z \log{\left(1 - a \right)} + 600 a^{2} z - 720 a^{2} \log{\left(1 - a \right)} + 1500 a p z \log{\left(1 - a \right)} - 540 a p z + 600 a p \log{\left(1 - a \right)} + 600 a z \log{\left(1 - a \right)} - 540 p z \log{\left(1 - a \right)}}{720 a^{4}}.$$

## Observer coefficient $J_{10}$

Set $a=1$, but replace the power with $Q(t)=t^3$ **before** integration. Substituting $p=1$ into $I_{10}$ would preserve linear power and is not this observer convention. The dimensional prefactor is the same.

In [3]:
OBSERVER_INTEGRAND = sp.cancel(INTEGRAND/Q).subs(a,1)*t**3
J10 = integrate_moments(OBSERVER_INTEGRAND, 0)
display(OBSERVER_INTEGRAND, J10)


t**2*(t**4*z - 3*t**3*z + t**3 + 3*t**2*z - 2*t**2 - t*z + t)/2

-(3*z - 7)/840

### Result

$$J_{10}=\frac{1}{120} - \frac{z}{280}.$$

## Correspondence with the existing interior source

The falling half after the source node has width $b=(\chi_{n+2}-\chi_{n+1})/\chi_{n+1}$. Its full-support contribution is $D(t,b)=b/2+t[1-(1+b)\log(1+b)/b]$. Therefore $D\to0$ as $b\to0^+$, and the interior source $H+D$ becomes $H$. The following also verifies the integrated result against the original NS B06 text export, rather than directly substituting into a singular $1/b$ expression.

In [4]:
import re
from pathlib import Path

D = b/2+t*(1-(1+b)*sp.log(1+b)/b)
assert sp.limit(D,b,0,dir='+') == 0
INTERIOR_INTEGRAND = (H+D)*(t-1+a)/(a*t)*Q*Z
assert sp.simplify(sp.limit(INTERIOR_INTEGRAND,b,0,dir='+')-INTEGRAND) == 0
FILENAME = 'Coefficient_B06.txt'
HERE = Path.cwd()
for CANDIDATE in (HERE, HERE/'notebooks'/'derivation'/'NS'):
    if (CANDIDATE/FILENAME).is_file():
        HERE = CANDIDATE
        break
ORIGINAL = (HERE/FILENAME).read_text()
PARTS = re.split(r'\b[IJ]\d+\s*=', ORIGINAL)[1:]
for TEXT, TARGET in zip(PARTS, (I10,J10), strict=True):
    EXPR = sp.sympify(TEXT.replace('numpy.log','log').replace('\n',' '), locals={'a':a,'b':b,'p':p,'z':z,'log':sp.log})
    # Denominator degree is <=2: terms through b^4 preserve its finite limit.
    REGULAR = sp.cancel(EXPR.subs(sp.log(1+b), b-b**2/2+b**3/3-b**4/4))
    assert sp.factor(REGULAR.subs(b,0)-TARGET) == 0
print('Both I and J agree with the original interior-source limit.')


Both I and J agree with the original interior-source limit.


## Stable evaluation and zero endpoint powers

Use $s=(1-t)/a\in[0,1]$ and $d_m=1/[(m+2)(m+3)]$. Define
$$f(v)=\sum_{m=0}^{\infty}d_m v^m,\qquad
g(v)=\sum_{m=0}^{\infty}\left(\frac12-\frac1{m+3}\right)v^m.$$
Then
$$F_a=a^2s^3f(as),\quad H_a=\frac{a^2s^2}2-(1-a)F_a,$$
$$\frac{F_a}{t}=a^2s^3g(as),\quad
\frac{H_a}{t}=\frac{a^2s^2}{2(1-as)}-(1-a)\frac{F_a}{t}.$$
For SS, $f(v)^2=\sum e_m v^m$, with $e_m=\sum_{j=0}^m d_jd_{m-j}$; multiply the two source expressions before integrating.

All remaining factors are polynomials in $s$:
$$P(s)=P_2+(P_1-P_2)s,\qquad (1+z)(s)=(1+z_2)+(z_1-z_2)s.$$
Integrate each resulting monomial using $\int_0^1s^qds=1/(q+1)$. This gives analytic series of **integrated moments**, with no production quadrature. The NS scale is $\chi_2a^3$, and the SS scale is $\chi_2^3a^5$ when endpoint powers/redshifts are included in the moments.

Production uses 16 terms for $a\le1/16$, 32 for $a\le1/4$, 64 for $a\le1/2$, and 160 for $a\le3/4$. The geometric tail parameter $a^K$ is below $6\times10^{-20}$ in each regime. The coefficients and moment factors are bounded, with at most a factor $1/(1-a)\le4$ from the geometric tail. Broader intervals use the displayed elementary closed form; the observer uses its separate $J$ expression. Floating-point cancellation from an intentionally sign-changing final spectrum is assessed with absolute error too.

The direct endpoint-linear moments avoid $P_1/P_2$. For the closed form, $I=A+Bp$ implies
$$P_2I=-BP_1+(A+B)P_2.$$
This equivalence handles either endpoint being zero without changing a signed contribution or introducing an epsilon.


In [5]:
LEFT_WEIGHT = sp.factor(-sp.diff(I10,p))
RIGHT_WEIGHT = sp.factor(I10.subs(p,0)-LEFT_WEIGHT)
assert sp.simplify(I10-(LEFT_WEIGHT*(1-p)+RIGHT_WEIGHT)) == 0
display(LEFT_WEIGHT, RIGHT_WEIGHT)


-(60*a**5*z*log(1 - a) - 113*a**5*z - 120*a**5*log(1 - a) + 250*a**5 - 60*a**4*z*log(1 - a) - 115*a**4*z + 120*a**4*log(1 - a) + 440*a**4 - 390*a**3*z + 1080*a**3*log(1 - a) - 1380*a**3 - 960*a**2*z*log(1 - a) + 1230*a**2*z - 1680*a**2*log(1 - a) + 600*a**2 + 1500*a*z*log(1 - a) - 540*a*z + 600*a*log(1 - a) - 540*z*log(1 - a))/(720*a**4)

-(60*a**5*z*log(1 - a) - 137*a**5*z - 240*a**5*log(1 - a) + 710*a**5 - 60*a**4*z*log(1 - a) - 325*a**4*z + 1680*a**4*log(1 - a) - 2240*a**4 - 1080*a**3*z*log(1 - a) + 1770*a**3*z - 3240*a**3*log(1 - a) + 2100*a**3 + 2640*a**2*z*log(1 - a) - 1830*a**2*z + 2400*a**2*log(1 - a) - 600*a**2 - 2100*a*z*log(1 - a) + 540*a*z - 600*a*log(1 - a) + 540*z*log(1 - a))/(720*a**4)

## Export and validation

[Coefficient_B10.txt](Coefficient_B10.txt) contains the ordinary and observer expressions in the existing `I10 = ...`, `J10 = ...` format. [Coefficient_B10_Validation.ipynb](Coefficient_B10_Validation.ipynb) compares the compiled float64 coefficient against independent integration of the original hats and moving source limits. The original Mathematica derivations remain unchanged.